crazy_rl/multi_agent/jax/base_parallel_env.py

```diff
@@ -0,0 +1,389 @@
+    glPushMatrix,
+    glShadeModel,
+)
+from OpenGL.raw.GLU import gluLookAt, gluPerspective
+from pettingzoo.utils.env import ParallelEnv
+from pygame import DOUBLEBUF, OPENGL
+
+from crazy_rl.utils.graphic import axes, field, point, target_point
+
+
+# from crazy_rl.utils.utils import run_land, run_sequence, run_take_off
+
+
+@jdc.pytree_dataclass
+class State:
+    """State of the environment containing the modifiable variables."""
+
+    timestep: int
+    agent_location: jnp.ndarray
+    terminations: jnp.ndarray
+    rewards: jnp.ndarray
+    observations: jnp.ndarray
+    truncations: jnp.ndarray
+
+
+class BaseParallelEnv(ParallelEnv):
+    """The Base environment inheriting from pettingZoo Parallel environment class.
+
+    The main API methods of this class are:
+    - step
+    - reset
+    - render
+    - close
+    - seed
+
+    they are defined in this main environment and the following attributes can be set in child env through the compute
+    method set:
+        action_space: The Space object corresponding to valid actions
+        observation_space: The Space object corresponding to valid observations
+        reward_range: A tuple corresponding to the min and max possible rewards
+    """
+
+    metadata = {
+        "render_modes": ["human", "real"],
+        "is_parallelizable": True,
+        "render_fps": 10,
+    }
+
+    def __init__(
+        self,
+        state: State,
+        num_drones: int,
+        init_flying_pos: Optional[jnp.ndarray] = None,
+        target_location: Optional[jnp.ndarray] = None,
+        size: int = 4,
+        render_mode: Optional[str] = None,
+        swarm: Optional[Swarm] = None,
+    ):
+        """Initialization of a generic aviary environment.
+
+        Args:
+            state: State containing modifiable variables of the environment
+            num_drones: ids of the drones (ignored in simulation mode)
+            init_flying_pos (array, optional): An array where each value is a (3)-shaped array containing the initial
+                XYZ position of the drones.
+            target_location (array, optional): An array containing a (3)-shaped array for the XYZ position of the target.
+            size (int, optional): Size of the area sides
+            render_mode (str, optional): The mode to display the rendering of the environment. Can be real, human or None.
+                Real mode is used for real tests on the field, human mode is used to display the environment on a PyGame
+                window and None mode is used to disable the rendering.
+            swarm (Swarm, optional): The Swarm object use in real mode to control all drones
+        """
+        # State initialisation
+        self.state = state
+
+        # Other initialisations
+        self.size = size  # The size of the square grid
+        self._init_flying_pos = init_flying_pos
+        self._target_location = target_location
+        self.num_drones = num_drones
+
+        self.nb_crash = 0
```

Comment:
These look like state variables or so?

Reply:
target_location is constant in surround so it's not in the state but it would be for escort and catch
nb_crash and nb_end are just for testing and I thought I would delete them at the end

Reply:
I put nb_crash and nb_end outside of the environment